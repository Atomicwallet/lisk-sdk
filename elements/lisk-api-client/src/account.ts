/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { codec } from '@liskhq/lisk-codec';
import { Channel, RegisteredSchemas } from './types';

export class Account {
	private readonly _channel: Channel;
	private readonly _schemas: RegisteredSchemas;

	public constructor(channel: Channel, schemas: RegisteredSchemas) {
		this._channel = channel;
		this._schemas = schemas;
	}

	public async get(address: Buffer): Promise<Record<string, unknown>> {
		const accountHex: string = await this._channel.invoke('app:getAccount', {
			address: address.toString('hex'),
		});

		return this.decode(Buffer.from(accountHex, 'hex'));
	}

	public encode(input: Record<string, unknown>): Buffer {
		return codec.encode(this._schemas.account, input);
	}

	public decode(input: Buffer): Record<string, unknown> {
		return codec.decode(this._schemas.account, input);
	}

	public toJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.toJSON(this._schemas.account, account);
	}

	public fromJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.fromJSON(this._schemas.account, account);
	}
}
