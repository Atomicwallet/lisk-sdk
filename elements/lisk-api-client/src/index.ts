/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export { createWSClient } from './ws_client';
export { createClient } from './client';
export { createAPIClient } from './create_api_client';
export { IPCChannel } from './ipc_channel';
