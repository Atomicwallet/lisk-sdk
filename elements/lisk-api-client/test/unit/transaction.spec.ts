/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { when } from 'jest-when';
import { getAddressAndPublicKeyFromPassphrase } from '@liskhq/lisk-cryptography';
import { Transaction } from '../../src/transaction';

describe('transaction', () => {
	let channelMock: any;
	let transaction: Transaction;
	const passphrases = ['trim elegant oven term access apple obtain error grain excite lawn neck'];
	const nodeInfo = {
		version:
			'3.0.0-beta.2.4-88b24e03bb28925a036293126dd96ac636218e29-6a1742532104af6f5c010e2ae77d3d982d471751.6a17425',
		networkVersion: '2.0',
		networkIdentifier: 'ccb837b25bc4f1b43fc08c2e80b07c6b46b84bf2264f6a37eaa4416fe478a0c5',
		lastBlockID: '57a669f7170239a68460af284eb1ba043839fe828be302b9c0e65fee498c954e',
		height: 276626,
		finalizedHeight: 276489,
		syncing: false,
		unconfirmedTransactions: 2243,
		genesisConfig: {
			blockTime: 10,
			maxPayloadLength: 15360,
			bftThreshold: 68,
			minFeePerByte: 1000,
			baseFees: [{ moduleID: 5, assetID: 0, baseFee: '1000000000' }],
			rewards: {
				milestones: ['500000000', '400000000', '300000000', '200000000', '100000000'],
				offset: 2160,
				distance: 3000000,
			},
			communityIdentifier: 'Lisk',
			minRemainingBalance: '5000000',
			activeDelegates: 101,
			standbyDelegates: 2,
			delegateListRoundOffset: 2,
		},
		registeredModules: [
			{
				id: 2,
				name: 'token',
				actions: [],
				events: [],
				reducers: [
					'token:credit',
					'token:debit',
					'token:getBalance',
					'token:getMinRemainingBalance',
				],
				transactionAssets: [{ id: 0, name: 'transfer' }],
			},
			{ id: 3, name: 'sequence', actions: [], events: [], reducers: [], transactionAssets: [] },
			{
				id: 4,
				name: 'keys',
				actions: [],
				events: [],
				reducers: [],
				transactionAssets: [{ id: 0, name: 'registerMultisignatureGroup' }],
			},
			{
				id: 5,
				name: 'dpos',
				actions: ['dpos:getAllDelegates'],
				events: [],
				reducers: [],
				transactionAssets: [
					{ id: 0, name: 'registerDelegate' },
					{ id: 1, name: 'voteDelegate' },
					{ id: 2, name: 'unlockToken' },
					{ id: 3, name: 'reportDelegateMisbehavior' },
				],
			},
			{
				id: 1000,
				name: 'legacyAccount',
				actions: [],
				events: [],
				reducers: [],
				transactionAssets: [{ id: 0, name: 'reclaimLSK' }],
			},
		],
	};
	const txHex =
		'0802100018362080ade2042a20dd4ff255fe04dd0159a468e9e9c8872c4f4466220f7e326377a0ceb9df2fa21a321d0880ade2041214654087c2df870402ab0b1996616fd3355d61f62c1a003a4079cb29dca7bb9fce73a1e8ca28264f779074d259c341b536bae9a54c0a2e4713580fcb192f9f15f43730650d69bb1f3dcfb4cb6da7d69ca990a763ed78569700';
	const accountHex =
		'0a14ab0041a7d3f7b2c290b5b834d46bdc7b7eb8581512050880c2d72f1a020800220208002a3b0a1a0a0a67656e657369735f3834180020850528003080a094a58d1d121d0a14ab0041a7d3f7b2c290b5b834d46bdc7b7eb858151080a094a58d1d';
	const encodedTx = Buffer.from(txHex, 'hex');
	const tx = {
		moduleID: 2,
		assetID: 0,
		nonce: BigInt('54'),
		fee: BigInt('10000000'),
		senderPublicKey: Buffer.from(
			'dd4ff255fe04dd0159a468e9e9c8872c4f4466220f7e326377a0ceb9df2fa21a',
			'hex',
		),
		asset: {
			amount: BigInt('10000000'),
			recipientAddress: Buffer.from('654087c2df870402ab0b1996616fd3355d61f62c', 'hex'),
			data: '',
		},
		signatures: [
			Buffer.from(
				'79cb29dca7bb9fce73a1e8ca28264f779074d259c341b536bae9a54c0a2e4713580fcb192f9f15f43730650d69bb1f3dcfb4cb6da7d69ca990a763ed78569700',
				'hex',
			),
		],
		id: 'dd93e4ca5b48d0b604e7cf2e57ce21be43a3163f853c83d88d383032fd830bbf',
	};
	const passphrase1 = 'trim elegant oven term access apple obtain error grain excite lawn neck';
	const { publicKey: publicKey1 } = getAddressAndPublicKeyFromPassphrase(passphrase1);
	const validTransaction = {
		moduleID: 2,
		assetID: 0,
		nonce: BigInt('1'),
		fee: BigInt('10000000'),
		senderPublicKey: publicKey1,
		asset: {
			recipientAddress: Buffer.from('3a971fd02b4a07fc20aad1936d3cb1d263b96e0f', 'hex'),
			amount: BigInt('4008489300000000'),
			data: '',
		},
	};
	const txId = Buffer.from(tx.id, 'hex');
	const schema = {
		account: {
			$id: '/account/base',
			type: 'object',
			properties: {
				address: { dataType: 'bytes', fieldNumber: 1 },
				token: {
					type: 'object',
					properties: { balance: { fieldNumber: 1, dataType: 'uint64' } },
					fieldNumber: 2,
				},
				sequence: {
					type: 'object',
					properties: { nonce: { fieldNumber: 1, dataType: 'uint64' } },
					fieldNumber: 3,
				},
				keys: {
					type: 'object',
					properties: {
						numberOfSignatures: { dataType: 'uint32', fieldNumber: 1 },
						mandatoryKeys: { type: 'array', items: { dataType: 'bytes' }, fieldNumber: 2 },
						optionalKeys: { type: 'array', items: { dataType: 'bytes' }, fieldNumber: 3 },
					},
					fieldNumber: 4,
				},
				dpos: {
					type: 'object',
					properties: {
						delegate: {
							type: 'object',
							fieldNumber: 1,
							properties: {
								username: { dataType: 'string', fieldNumber: 1 },
								pomHeights: { type: 'array', items: { dataType: 'uint32' }, fieldNumber: 2 },
								consecutiveMissedBlocks: { dataType: 'uint32', fieldNumber: 3 },
								lastForgedHeight: { dataType: 'uint32', fieldNumber: 4 },
								isBanned: { dataType: 'boolean', fieldNumber: 5 },
								totalVotesReceived: { dataType: 'uint64', fieldNumber: 6 },
							},
							required: [
								'username',
								'pomHeights',
								'consecutiveMissedBlocks',
								'lastForgedHeight',
								'isBanned',
								'totalVotesReceived',
							],
						},
						sentVotes: {
							type: 'array',
							fieldNumber: 2,
							items: {
								type: 'object',
								properties: {
									delegateAddress: { dataType: 'bytes', fieldNumber: 1 },
									amount: { dataType: 'uint64', fieldNumber: 2 },
								},
								required: ['delegateAddress', 'amount'],
							},
						},
						unlocking: {
							type: 'array',
							fieldNumber: 3,
							items: {
								type: 'object',
								properties: {
									delegateAddress: { dataType: 'bytes', fieldNumber: 1 },
									amount: { dataType: 'uint64', fieldNumber: 2 },
									unvoteHeight: { dataType: 'uint32', fieldNumber: 3 },
								},
								required: ['delegateAddress', 'amount', 'unvoteHeight'],
							},
						},
					},
					fieldNumber: 5,
				},
			},
			required: ['address', 'token', 'sequence', 'keys', 'dpos'],
		},
		block: {
			$id: '/block',
			type: 'object',
			properties: {
				header: { dataType: 'bytes', fieldNumber: 1 },
				payload: { type: 'array', items: { dataType: 'bytes' }, fieldNumber: 2 },
			},
			required: ['header', 'payload'],
		},
		blockHeader: {
			$id: '/block/header',
			type: 'object',
			properties: {
				version: { dataType: 'uint32', fieldNumber: 1 },
				timestamp: { dataType: 'uint32', fieldNumber: 2 },
				height: { dataType: 'uint32', fieldNumber: 3 },
				previousBlockID: { dataType: 'bytes', fieldNumber: 4 },
				transactionRoot: { dataType: 'bytes', fieldNumber: 5 },
				generatorPublicKey: { dataType: 'bytes', fieldNumber: 6 },
				reward: { dataType: 'uint64', fieldNumber: 7 },
				asset: { dataType: 'bytes', fieldNumber: 8 },
				signature: { dataType: 'bytes', fieldNumber: 9 },
			},
			required: [
				'version',
				'timestamp',
				'height',
				'previousBlockID',
				'transactionRoot',
				'generatorPublicKey',
				'reward',
				'asset',
			],
		},
		blockHeadersAssets: {
			0: {
				$id: '/genesisBlock/header/asset',
				type: 'object',
				required: ['accounts', 'initDelegates', 'initRounds'],
				properties: {
					accounts: {
						type: 'array',
						fieldNumber: 1,
						items: {
							$id: '/account/base',
							type: 'object',
							properties: {
								address: { dataType: 'bytes', fieldNumber: 1 },
								token: {
									type: 'object',
									properties: { balance: { fieldNumber: 1, dataType: 'uint64' } },
									fieldNumber: 2,
								},
								sequence: {
									type: 'object',
									properties: { nonce: { fieldNumber: 1, dataType: 'uint64' } },
									fieldNumber: 3,
								},
								keys: {
									type: 'object',
									properties: {
										numberOfSignatures: { dataType: 'uint32', fieldNumber: 1 },
										mandatoryKeys: { type: 'array', items: { dataType: 'bytes' }, fieldNumber: 2 },
										optionalKeys: { type: 'array', items: { dataType: 'bytes' }, fieldNumber: 3 },
									},
									fieldNumber: 4,
								},
								dpos: {
									type: 'object',
									properties: {
										delegate: {
											type: 'object',
											fieldNumber: 1,
											properties: {
												username: { dataType: 'string', fieldNumber: 1 },
												pomHeights: {
													type: 'array',
													items: { dataType: 'uint32' },
													fieldNumber: 2,
												},
												consecutiveMissedBlocks: { dataType: 'uint32', fieldNumber: 3 },
												lastForgedHeight: { dataType: 'uint32', fieldNumber: 4 },
												isBanned: { dataType: 'boolean', fieldNumber: 5 },
												totalVotesReceived: { dataType: 'uint64', fieldNumber: 6 },
											},
											required: [
												'username',
												'pomHeights',
												'consecutiveMissedBlocks',
												'lastForgedHeight',
												'isBanned',
												'totalVotesReceived',
											],
										},
										sentVotes: {
											type: 'array',
											fieldNumber: 2,
											items: {
												type: 'object',
												properties: {
													delegateAddress: { dataType: 'bytes', fieldNumber: 1 },
													amount: { dataType: 'uint64', fieldNumber: 2 },
												},
												required: ['delegateAddress', 'amount'],
											},
										},
										unlocking: {
											type: 'array',
											fieldNumber: 3,
											items: {
												type: 'object',
												properties: {
													delegateAddress: { dataType: 'bytes', fieldNumber: 1 },
													amount: { dataType: 'uint64', fieldNumber: 2 },
													unvoteHeight: { dataType: 'uint32', fieldNumber: 3 },
												},
												required: ['delegateAddress', 'amount', 'unvoteHeight'],
											},
										},
									},
									fieldNumber: 5,
								},
							},
							required: ['address', 'token', 'sequence', 'keys', 'dpos'],
						},
					},
					initDelegates: {
						type: 'array',
						items: { dataType: 'bytes' },
						fieldNumber: 2,
						minItems: 1,
					},
					initRounds: { dataType: 'uint32', fieldNumber: 3, minimum: 3 },
				},
			},
			2: {
				$id: '/blockHeader/asset/v2',
				type: 'object',
				properties: {
					maxHeightPreviouslyForged: { dataType: 'uint32', fieldNumber: 1 },
					maxHeightPrevoted: { dataType: 'uint32', fieldNumber: 2 },
					seedReveal: { dataType: 'bytes', minLength: 16, maxLength: 16, fieldNumber: 3 },
				},
				required: ['maxHeightPreviouslyForged', 'maxHeightPrevoted', 'seedReveal'],
			},
		},
		transaction: {
			$id: 'lisk/transaction',
			type: 'object',
			required: ['moduleID', 'assetID', 'nonce', 'fee', 'senderPublicKey', 'asset'],
			properties: {
				moduleID: { dataType: 'uint32', fieldNumber: 1, minimum: 2 },
				assetID: { dataType: 'uint32', fieldNumber: 2 },
				nonce: { dataType: 'uint64', fieldNumber: 3 },
				fee: { dataType: 'uint64', fieldNumber: 4 },
				senderPublicKey: { dataType: 'bytes', fieldNumber: 5, minLength: 32, maxLength: 32 },
				asset: { dataType: 'bytes', fieldNumber: 6 },
				signatures: { type: 'array', items: { dataType: 'bytes' }, fieldNumber: 7 },
			},
		},
		transactionsAssets: [
			{
				moduleID: 2,
				moduleName: 'token',
				assetID: 0,
				assetName: 'transfer',
				schema: {
					$id: 'lisk/transfer-asset',
					title: 'Transfer transaction asset',
					type: 'object',
					required: ['amount', 'recipientAddress', 'data'],
					properties: {
						amount: { dataType: 'uint64', fieldNumber: 1 },
						recipientAddress: { dataType: 'bytes', fieldNumber: 2, minLength: 20, maxLength: 20 },
						data: { dataType: 'string', fieldNumber: 3, minLength: 0, maxLength: 64 },
					},
				},
			},
			{
				moduleID: 4,
				moduleName: 'keys',
				assetID: 0,
				assetName: 'registerMultisignatureGroup',
				schema: {
					$id: 'lisk/keys/register',
					type: 'object',
					required: ['numberOfSignatures', 'optionalKeys', 'mandatoryKeys'],
					properties: {
						numberOfSignatures: { dataType: 'uint32', fieldNumber: 1, minimum: 1, maximum: 64 },
						mandatoryKeys: {
							type: 'array',
							items: { dataType: 'bytes', minLength: 32, maxLength: 32 },
							fieldNumber: 2,
							minItems: 0,
							maxItems: 64,
						},
						optionalKeys: {
							type: 'array',
							items: { dataType: 'bytes', minLength: 32, maxLength: 32 },
							fieldNumber: 3,
							minItems: 0,
							maxItems: 64,
						},
					},
				},
			},
			{
				moduleID: 5,
				moduleName: 'dpos',
				assetID: 0,
				assetName: 'registerDelegate',
				schema: {
					$id: 'lisk/dpos/register',
					type: 'object',
					required: ['username'],
					properties: {
						username: { dataType: 'string', fieldNumber: 1, minLength: 1, maxLength: 20 },
					},
				},
			},
			{
				moduleID: 5,
				moduleName: 'dpos',
				assetID: 1,
				assetName: 'voteDelegate',
				schema: {
					$id: 'lisk/dpos/vote',
					type: 'object',
					required: ['votes'],
					properties: {
						votes: {
							type: 'array',
							minItems: 1,
							maxItems: 20,
							items: {
								type: 'object',
								required: ['delegateAddress', 'amount'],
								properties: {
									delegateAddress: {
										dataType: 'bytes',
										fieldNumber: 1,
										minLength: 20,
										maxLength: 20,
									},
									amount: { dataType: 'sint64', fieldNumber: 2 },
								},
							},
							fieldNumber: 1,
						},
					},
				},
			},
			{
				moduleID: 5,
				moduleName: 'dpos',
				assetID: 2,
				assetName: 'unlockToken',
				schema: {
					$id: 'lisk/dpos/unlock',
					type: 'object',
					required: ['unlockObjects'],
					properties: {
						unlockObjects: {
							type: 'array',
							minItems: 1,
							maxItems: 20,
							items: {
								type: 'object',
								required: ['delegateAddress', 'amount', 'unvoteHeight'],
								properties: {
									delegateAddress: {
										dataType: 'bytes',
										fieldNumber: 1,
										minLength: 20,
										maxLength: 20,
									},
									amount: { dataType: 'uint64', fieldNumber: 2 },
									unvoteHeight: { dataType: 'uint32', fieldNumber: 3 },
								},
							},
							fieldNumber: 1,
						},
					},
				},
			},
			{
				moduleID: 5,
				moduleName: 'dpos',
				assetID: 3,
				assetName: 'reportDelegateMisbehavior',
				schema: {
					$id: 'lisk/dpos/pom',
					type: 'object',
					required: ['header1', 'header2'],
					properties: {
						header1: {
							$id: 'lisk/block-header',
							type: 'object',
							properties: {
								version: { dataType: 'uint32', fieldNumber: 1 },
								timestamp: { dataType: 'uint32', fieldNumber: 2 },
								height: { dataType: 'uint32', fieldNumber: 3 },
								previousBlockID: { dataType: 'bytes', fieldNumber: 4 },
								transactionRoot: { dataType: 'bytes', fieldNumber: 5 },
								generatorPublicKey: { dataType: 'bytes', fieldNumber: 6 },
								reward: { dataType: 'uint64', fieldNumber: 7 },
								asset: {
									type: 'object',
									fieldNumber: 8,
									properties: {
										maxHeightPreviouslyForged: { dataType: 'uint32', fieldNumber: 1 },
										maxHeightPrevoted: { dataType: 'uint32', fieldNumber: 2 },
										seedReveal: { dataType: 'bytes', fieldNumber: 3 },
									},
									required: ['maxHeightPreviouslyForged', 'maxHeightPrevoted', 'seedReveal'],
								},
								signature: { dataType: 'bytes', fieldNumber: 9 },
							},
							required: [
								'version',
								'timestamp',
								'height',
								'previousBlockID',
								'transactionRoot',
								'generatorPublicKey',
								'reward',
								'asset',
							],
							fieldNumber: 1,
						},
						header2: {
							$id: 'lisk/block-header',
							type: 'object',
							properties: {
								version: { dataType: 'uint32', fieldNumber: 1 },
								timestamp: { dataType: 'uint32', fieldNumber: 2 },
								height: { dataType: 'uint32', fieldNumber: 3 },
								previousBlockID: { dataType: 'bytes', fieldNumber: 4 },
								transactionRoot: { dataType: 'bytes', fieldNumber: 5 },
								generatorPublicKey: { dataType: 'bytes', fieldNumber: 6 },
								reward: { dataType: 'uint64', fieldNumber: 7 },
								asset: {
									type: 'object',
									fieldNumber: 8,
									properties: {
										maxHeightPreviouslyForged: { dataType: 'uint32', fieldNumber: 1 },
										maxHeightPrevoted: { dataType: 'uint32', fieldNumber: 2 },
										seedReveal: { dataType: 'bytes', fieldNumber: 3 },
									},
									required: ['maxHeightPreviouslyForged', 'maxHeightPrevoted', 'seedReveal'],
								},
								signature: { dataType: 'bytes', fieldNumber: 9 },
							},
							required: [
								'version',
								'timestamp',
								'height',
								'previousBlockID',
								'transactionRoot',
								'generatorPublicKey',
								'reward',
								'asset',
							],
							fieldNumber: 2,
						},
					},
				},
			},
			{
				moduleID: 1000,
				moduleName: 'legacyAccount',
				assetID: 0,
				assetName: 'reclaimLSK',
				schema: {
					$id: 'lisk/legacyAccount/reclaim',
					title: 'Reclaim transaction asset',
					type: 'object',
					required: ['amount'],
					properties: { amount: { dataType: 'uint64', fieldNumber: 1 } },
				},
			},
		],
	};

	beforeEach(() => {
		channelMock = {
			connect: jest.fn(),
			disconnect: jest.fn(),
			invoke: jest.fn(),
			subscribe: jest.fn(),
		};

		when(channelMock.invoke)
			.calledWith('app:getAccount')
			.mockResolvedValue(accountHex as never)
			.calledWith('app:getTransactionByID')
			.mockResolvedValue(txHex as never)
			.calledWith('app:getTransactionsFromPool')
			.mockResolvedValue(txHex as never);

		transaction = new Transaction(channelMock, schema, nodeInfo);
	});

	describe('Transaction', () => {
		describe('constructor', () => {
			it('should initialize with channel', () => {
				expect(transaction['_channel']).toBe(channelMock);
			});
		});

		describe('get', () => {
			it('should invoke app:getTransactionByID', async () => {
				// Act
				await transaction.get(txId);

				// Assert
				expect(channelMock.invoke).toHaveBeenCalledTimes(1);
				expect(channelMock.invoke).toHaveBeenCalledWith('app:getTransactionByID', {
					id: txId.toString('hex'),
				});
			});
		});

		describe('getFromPool', () => {
			it('should invoke app:getTransactionsFromPool', async () => {
				// Act
				await transaction.getFromPool();

				// Assert
				expect(channelMock.invoke).toHaveBeenCalledTimes(1);
				expect(channelMock.invoke).toHaveBeenCalledWith('app:getTransactionsFromPool');
			});
		});

		describe('create', () => {
			it('should return created tx', async () => {
				// Act
				const returnedTx = await transaction.create(validTransaction, passphrase1);

				// Assert
				expect(returnedTx.signatures).toHaveLength(1);
				expect(returnedTx.signatures).toMatchSnapshot();
			});
		});

		describe('sign', () => {
			it('should return some signed transaction', () => {
				// Act
				const returnedTx = transaction.sign(validTransaction, passphrases);

				// Assert
				expect(returnedTx).toBeDefined();
			});
		});

		describe('send', () => {
			it('should invoke app:postTransaction', async () => {
				// Act
				await transaction.send(tx);

				// Assert
				expect(channelMock.invoke).toHaveBeenCalledTimes(1);
				expect(channelMock.invoke).toHaveBeenCalledWith('app:postTransaction', {
					transaction: txHex,
				});
			});
		});

		describe('decode', () => {
			it('should return decoded transaction', () => {
				// Act
				const decodedTx = transaction.decode(encodedTx);

				// Assert
				expect(decodedTx).toEqual(tx);
			});
		});

		describe('encode', () => {
			it('should return encoded transaction', () => {
				// Act
				const returnedTx = transaction.encode(tx);

				// Assert
				expect(returnedTx).toEqual(encodedTx);
			});
		});

		describe('getMinFee', () => {
			it('should return some value', () => {
				// Act
				const fee = transaction.getMinFee(tx);

				// Assert
				expect(fee).toBeDefined();
			});
		});
	});
});
