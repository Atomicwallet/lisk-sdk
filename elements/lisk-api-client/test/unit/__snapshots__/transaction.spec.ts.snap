// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction Transaction create when called with a valid transaction should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      142,
      63,
      76,
      225,
      230,
      221,
      48,
      163,
      28,
      32,
      167,
      106,
      182,
      106,
      33,
      91,
      114,
      232,
      242,
      20,
      48,
      137,
      145,
      28,
      128,
      59,
      137,
      244,
      212,
      49,
      254,
      78,
      20,
      106,
      134,
      66,
      181,
      59,
      58,
      161,
      16,
      31,
      185,
      148,
      133,
      124,
      247,
      147,
      126,
      143,
      44,
      111,
      153,
      237,
      131,
      96,
      219,
      134,
      254,
      38,
      43,
      135,
      158,
      11,
    ],
    "type": "Buffer",
  },
]
`;

exports[`transaction Transaction create when called with multi-signature account in input should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      142,
      63,
      76,
      225,
      230,
      221,
      48,
      163,
      28,
      32,
      167,
      106,
      182,
      106,
      33,
      91,
      114,
      232,
      242,
      20,
      48,
      137,
      145,
      28,
      128,
      59,
      137,
      244,
      212,
      49,
      254,
      78,
      20,
      106,
      134,
      66,
      181,
      59,
      58,
      161,
      16,
      31,
      185,
      148,
      133,
      124,
      247,
      147,
      126,
      143,
      44,
      111,
      153,
      237,
      131,
      96,
      219,
      134,
      254,
      38,
      43,
      135,
      158,
      11,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [],
    "type": "Buffer",
  },
]
`;

exports[`transaction Transaction create when called with nonce equal to zero in input should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      213,
      41,
      81,
      89,
      78,
      131,
      114,
      241,
      106,
      63,
      52,
      237,
      204,
      195,
      225,
      154,
      195,
      189,
      130,
      20,
      196,
      204,
      121,
      218,
      187,
      200,
      176,
      156,
      150,
      202,
      199,
      45,
      233,
      62,
      6,
      243,
      49,
      50,
      144,
      109,
      178,
      84,
      153,
      244,
      163,
      237,
      103,
      35,
      131,
      176,
      203,
      85,
      31,
      146,
      226,
      139,
      188,
      59,
      242,
      122,
      129,
      28,
      164,
      15,
    ],
    "type": "Buffer",
  },
]
`;

exports[`transaction Transaction create when called with optional keys in input should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      142,
      63,
      76,
      225,
      230,
      221,
      48,
      163,
      28,
      32,
      167,
      106,
      182,
      106,
      33,
      91,
      114,
      232,
      242,
      20,
      48,
      137,
      145,
      28,
      128,
      59,
      137,
      244,
      212,
      49,
      254,
      78,
      20,
      106,
      134,
      66,
      181,
      59,
      58,
      161,
      16,
      31,
      185,
      148,
      133,
      124,
      247,
      147,
      126,
      143,
      44,
      111,
      153,
      237,
      131,
      96,
      219,
      134,
      254,
      38,
      43,
      135,
      158,
      11,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [],
    "type": "Buffer",
  },
]
`;

exports[`transaction Transaction create when called without sender public key in input should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      142,
      63,
      76,
      225,
      230,
      221,
      48,
      163,
      28,
      32,
      167,
      106,
      182,
      106,
      33,
      91,
      114,
      232,
      242,
      20,
      48,
      137,
      145,
      28,
      128,
      59,
      137,
      244,
      212,
      49,
      254,
      78,
      20,
      106,
      134,
      66,
      181,
      59,
      58,
      161,
      16,
      31,
      185,
      148,
      133,
      124,
      247,
      147,
      126,
      143,
      44,
      111,
      153,
      237,
      131,
      96,
      219,
      134,
      254,
      38,
      43,
      135,
      158,
      11,
    ],
    "type": "Buffer",
  },
]
`;

exports[`transaction Transaction decode should return decoded transaction 1`] = `
Object {
  "asset": Object {
    "amount": 10000000n,
    "data": "",
    "recipientAddress": Object {
      "data": Array [
        101,
        64,
        135,
        194,
        223,
        135,
        4,
        2,
        171,
        11,
        25,
        150,
        97,
        111,
        211,
        53,
        93,
        97,
        246,
        44,
      ],
      "type": "Buffer",
    },
  },
  "assetID": 0,
  "fee": 10000000n,
  "id": Object {
    "data": Array [
      221,
      147,
      228,
      202,
      91,
      72,
      208,
      182,
      4,
      231,
      207,
      46,
      87,
      206,
      33,
      190,
      67,
      163,
      22,
      63,
      133,
      60,
      131,
      216,
      141,
      56,
      48,
      50,
      253,
      131,
      11,
      191,
    ],
    "type": "Buffer",
  },
  "moduleID": 2,
  "nonce": 54n,
  "senderPublicKey": Object {
    "data": Array [
      221,
      79,
      242,
      85,
      254,
      4,
      221,
      1,
      89,
      164,
      104,
      233,
      233,
      200,
      135,
      44,
      79,
      68,
      102,
      34,
      15,
      126,
      50,
      99,
      119,
      160,
      206,
      185,
      223,
      47,
      162,
      26,
    ],
    "type": "Buffer",
  },
  "signatures": Array [
    Object {
      "data": Array [
        121,
        203,
        41,
        220,
        167,
        187,
        159,
        206,
        115,
        161,
        232,
        202,
        40,
        38,
        79,
        119,
        144,
        116,
        210,
        89,
        195,
        65,
        181,
        54,
        186,
        233,
        165,
        76,
        10,
        46,
        71,
        19,
        88,
        15,
        203,
        25,
        47,
        159,
        21,
        244,
        55,
        48,
        101,
        13,
        105,
        187,
        31,
        61,
        207,
        180,
        203,
        109,
        167,
        214,
        156,
        169,
        144,
        167,
        99,
        237,
        120,
        86,
        151,
        0,
      ],
      "type": "Buffer",
    },
  ],
}
`;

exports[`transaction Transaction sign when called with multi-signature account in input should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      142,
      63,
      76,
      225,
      230,
      221,
      48,
      163,
      28,
      32,
      167,
      106,
      182,
      106,
      33,
      91,
      114,
      232,
      242,
      20,
      48,
      137,
      145,
      28,
      128,
      59,
      137,
      244,
      212,
      49,
      254,
      78,
      20,
      106,
      134,
      66,
      181,
      59,
      58,
      161,
      16,
      31,
      185,
      148,
      133,
      124,
      247,
      147,
      126,
      143,
      44,
      111,
      153,
      237,
      131,
      96,
      219,
      134,
      254,
      38,
      43,
      135,
      158,
      11,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      125,
      251,
      192,
      234,
      222,
      176,
      53,
      65,
      207,
      190,
      10,
      123,
      252,
      141,
      191,
      61,
      31,
      7,
      188,
      35,
      215,
      5,
      81,
      28,
      56,
      105,
      12,
      1,
      6,
      236,
      137,
      186,
      96,
      52,
      171,
      169,
      172,
      161,
      52,
      227,
      37,
      236,
      32,
      8,
      138,
      89,
      24,
      185,
      190,
      19,
      175,
      185,
      136,
      14,
      5,
      226,
      121,
      21,
      15,
      150,
      145,
      11,
      218,
      8,
    ],
    "type": "Buffer",
  },
]
`;

exports[`transaction Transaction sign when called with optional keys in input should return created tx 1`] = `
Array [
  Object {
    "data": Array [
      142,
      63,
      76,
      225,
      230,
      221,
      48,
      163,
      28,
      32,
      167,
      106,
      182,
      106,
      33,
      91,
      114,
      232,
      242,
      20,
      48,
      137,
      145,
      28,
      128,
      59,
      137,
      244,
      212,
      49,
      254,
      78,
      20,
      106,
      134,
      66,
      181,
      59,
      58,
      161,
      16,
      31,
      185,
      148,
      133,
      124,
      247,
      147,
      126,
      143,
      44,
      111,
      153,
      237,
      131,
      96,
      219,
      134,
      254,
      38,
      43,
      135,
      158,
      11,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      125,
      251,
      192,
      234,
      222,
      176,
      53,
      65,
      207,
      190,
      10,
      123,
      252,
      141,
      191,
      61,
      31,
      7,
      188,
      35,
      215,
      5,
      81,
      28,
      56,
      105,
      12,
      1,
      6,
      236,
      137,
      186,
      96,
      52,
      171,
      169,
      172,
      161,
      52,
      227,
      37,
      236,
      32,
      8,
      138,
      89,
      24,
      185,
      190,
      19,
      175,
      185,
      136,
      14,
      5,
      226,
      121,
      21,
      15,
      150,
      145,
      11,
      218,
      8,
    ],
    "type": "Buffer",
  },
]
`;
